File: cpp_solutions/1052.cpp
class Solution {
public:
    int maxSatisfied(vector<int>& customers, vector<int>& grumpy, int minutes) {
        int base = 0;
        int n = customers.size();
        for (int i = 0; i < n; ++i) {
            if (grumpy[i] == 0) base += customers[i];
        }
        int left = 0;
        int plus = 0;
        int res = base;
        for (int right = 0; right < n; ++right) {
            if (grumpy[right] == 1) plus += customers[right];
            while (right - left + 1 > minutes) {
                if (grumpy[left] == 1) plus -= customers[left];
                left++;
            }
            res = max(res, base + plus);
        }
        return res;
    }
};

File: rust_solutions/1052.rs
impl Solution {
    pub fn max_satisfied(customers: Vec<i32>, grumpy: Vec<i32>, minutes: i32) -> i32 {
        let n = customers.len() as i32;
        let mut accum = 0;
        for i in 0..n {
            if grumpy[i as usize] == 0 {
                accum += customers[i as usize];
            }
        }

        let mut maxSatisfied = accum;
        for i in 0..minutes - 1 {
            if grumpy[i as usize] == 1 {
                accum += customers[i as usize];
            }
        }

        for i in minutes - 1..n {
            if grumpy[i as usize] == 1 {
                accum += customers[i as usize];
            }
            maxSatisfied = i32::max(maxSatisfied, accum);
            if grumpy[(i - minutes + 1) as usize] == 1 {
                accum -= customers[(i - minutes + 1) as usize];
            }
        }
        maxSatisfied
    }
}

